reporter_test.go

```diff
@@ -465,7 +496,115 @@ func TestDiffFile(t *testing.T) {
 				t.Errorf("diffFile() error: %v", err)
 			default:
 				if gotDiff != tc.wantDiff {
-					t.Errorf("diffFile() diff: got=%s, want=%s", gotDiff, tc.wantDiff)
+					t.Errorf("diffFile() diff: got=%q, want=%q", gotDiff, tc.wantDiff)
+				}
+			}
+		})
+	}
+}
+
+func TestCompare(t *testing.T) {
+	testCases := []struct {
+		desc      string
+		before    *fspb.Walk
+		after     *fspb.Walk
+		deleted   int
+		added     int
+		modified  int
+		wantError bool
+	}{
+		{
+			desc:   "nil before",
+			before: nil,
+			after: &fspb.Walk{
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+				},
+			},
+			added: 1,
+		}, {
+			desc: "empty after",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+				},
+			},
+			after:   &fspb.Walk{Id: "2"},
+			deleted: 1,
+		}, {
+			desc:      "nil before and after",
+			before:    nil,
+			after:     nil,
+			wantError: true,
+		}, {
+			desc: "diffs",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+					&fspb.File{Path: "/e/f/g", Info: &fspb.FileInfo{Size: 4}},
+					&fspb.File{Path: "/x/y/z", Info: &fspb.FileInfo{}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/b/c/d", Info: &fspb.FileInfo{}},
+					&fspb.File{Path: "/e/f/g", Info: &fspb.FileInfo{Size: 7}},
+					&fspb.File{Path: "/x/y/z", Info: &fspb.FileInfo{}},
+				},
+			},
+			added:    1,
+			deleted:  1,
+			modified: 1,
+		}, {
+			desc: "ignore",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/ignore/a", Info: &fspb.FileInfo{}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/ignore/b", Info: &fspb.FileInfo{}},
+				},
+			},
+		}, {
+			desc: "same dir with and without trailing /",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c/", Info: &fspb.FileInfo{IsDir: true}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{IsDir: true}},
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			r := &Reporter{config: &fspb.ReportConfig{ExcludePfx: []string{"/ignore/"}}}
+			report, err := r.Compare(tc.before, tc.after)
+			if tc.wantError {
+				if err == nil {
+					t.Error("Compare() no error")
+				}
+			} else {
```

Comment:
Ah but what happens now when `tc.wantError && err == nil`? You'll have to either add another check in this `else` branch or, what I often prefer personally, is use a switch:

```go
swtich {
case tc.wantError && err == nil:
  t.Error("r.Compare returned nil error")
case !tc.wantError && err != nil:
  t.Errorf("r.Compare: %v", err)
case err == nil:
  // check len(report.Added), Deleted and Modified here
}
```

Reply:
Ehm, sorry I meant what happens when `!tc.wantError && err != nil`